### ISSUE TITLE
Can I write a thread_safe python user defined problem?

### ISSUE THREAD

**Issue by chengdi123000** [esa/pagmo2#227]
I cannot get a problem even I write a `get_thread_safety(self)` function to return `True`.

I want to know if this is a limitation of python API.

**Comment by chengdi123000** [comment]
Thank you for your reply. 

I want it because threads can share a lot of things. Message passing between processes is so painful. 

I have a model fitting problem with about 200 experiment data point. I want to use some optimization method to find the pareto front of two objectives, which are fraction of variance unexplained (FVU)s of two performance index. The model simulation invokes a lot of complex cython objects. So it is difficult to do serialization.

I have two choices: 
1. distribute simulation of every data point to ipyparallel nodes. FVU are calculated in master node
2. distribute FVU calculation to ipyparallel engines. Master node create an island to hold the status.

I prefer the first one, but I am confusing. There are two additional level of parallelism in pagmo2: islands and archipelagos. And pickle is also a source of many errors. I have to define some `__reduce__` method to transfer the user defined problem class.

**Comment by bluescarni** [comment]
At the moment we don't support parallelisation via threads in python.

Note that, as far as CPython is concerned, you can get parallel speedup with multiple threads only if you are manually releasing the GIL whenever you are not using the Python interpreter from the current thread (and of course the computationally intensive parts of your code have to be written in some language other than Python). Are you doing this type of GIL management bits by yourself (or perhaps is Cython is doing it for you)? Because if you are not, you will not get any speedup from multithreaded Python.

You don't have to use ipyparallel necessarily, unless you are on Python 2. If a single computer is enough for your optimisation tasks, then you can use the multiprocessing island (``mp_island``). I believe it is possible to share at least certain data types between processes via shared memory in python, so perhaps if you are having troubles with serialization you could use such facilities to communicate between your multiprocessing islands. See for instance this SO answer:

https://stackoverflow.com/questions/10721915/shared-memory-objects-in-multiprocessing

**Comment by chengdi123000** [comment]
Unfortunately, my program may need more nodes for memory and FLOPs. So ipyparallel is needed. I just implement a `__reduce__` and use `Client.use_pickle` method to transfer the state of problem. It works. The default enhanced pickle library of ipyparallel `cloudpickle` have some trouble with cython objects even when I write a `__reduce__` for user defined problem.

**Comment by chengdi123000** [comment]
But I found that `ipyparallel_island` do not work well with my method because it must use `cloudpickle` to do serialization.

**Comment by bluescarni** [comment]
We are using ``cloudpickle`` by default across pagmo because it allows us to pickle objects which the standard ``pickle`` module is unable to serialize (most importantly, objects defined in interactive sessions, lambda functions and the likes).

With that said, it would probably be good to give the user the choice of the serialization backend, as unfortunately ``cloudpickle`` does not seem to satisfy all use cases. What kind of errors are you getting with cloudpickle? Should these be reported upstream?

**Comment by chengdi123000** [comment]
@bluescarni 
When I define a function with some cython objects which is not naturally serializable.  `pickle` just dumps the name of the function, while `cloudpickle` tends to dumps everything related to it and report missing `__reduce__` method of the cython object. 

This is not directly related to the `ipyparallel_island`.

**Comment by bluescarni** [comment]
I took a look yesterday and it should not be too hard for me to add an option to switch serialization backend in pygmo. I can take a stab at it.

Are you compiling pygmo from source or using a pre-built package? This change will not be available in the binary packages until the next pygmo version, which may still take some time, so it would be best if you could test it out from the current git head (once I have implemented it, that is).

**Comment by chengdi123000** [comment]
I am using pip installed pre-built package. 

My coding pattern is to get a ipyparallel.Client, then use the `%%px --local ` cell magic to do some initialization.

If possible, can we just follow the Client's setting? In ipyparallel, there is `use_pickle()` and `use_cloudpickle` function. Or can I transfer DirectView or Client object as an argument of constructor?

**Comment by bluescarni** [comment]
Do you have a small (ideally minimal) example of serialization failure with the current pygmo version that you could provide here (something based on cython perhaps)?

I am reworking the serialization bits to allow the selection of the serialization backend, and it would be good to have an actual example of serialization failure that we can now fix with this new machinery.

I am planning to add cython-based tests to the unit tests in order to make sure that this stays fixed, once we fix it.

**Comment by bluescarni** [comment]
Regarding your question about the ipyparallel client settings: the ipyparallel island allows you to specify whatever parameter you want for the construction of the ipyparallel client that will be used. Quoting the documentation (https://esa.github.io/pagmo2/docs/python/islands/py_islands.html#pygmo.ipyparallel_island):

> The arguments to the constructor of this class will be passed without modifications to the constructor of the ipyparallel.Client object that will be created internally.

So if you construct an ipyparallel_island with construction parameters ``(a, b, c)``, then these parameters will be passed to the constructor of the ipyparallel client.

It is true that we enforce the use of ``cloudpickle`` in the initial ipyparallel setup, but I realised that those calls to ``use_cloudpickle()`` are actually not doing anything, as the serialization of UDPs/UDAs etc. ignores this setting. So I have just removed the call to ``use_cloudpickle()`` in #229 to avoid confusion.
